File: patterns/structural_decorator/structural_decorator_webref.js
class Car {
    constructor() {
        this.price = 10000;
        this.model = 'Car';
    }

    getPrice() {
        return this.price;
    }

    getDescription() {
        return this.model;
    }
}

class Tesla extends Car {
    constructor() {
        super();
        this.price = 25000;
        this.model = 'Tesla';
    }
}

// декораторы

class Autopilot {
    constructor(car) {
        this.car = car;
    }

    getPrice() {
        return this.car.getPrice() + 5000;
    }

    getDescription() {
        return `${this.car.getDescription()} with autopilot`;
    }
};

class Parktronic {
    constructor(car) {
        this.car = car;
    }

    getPrice() {
        return this.car.getPrice() + 3000;
    }

    getDescription() {
        return `${this.car.getDescription()} with parktronic`;
    }
};


// ---
// +autopilot + parktronic

let tesla = new Tesla();
tesla = new Autopilot(tesla);
tesla = new Parktronic(tesla);

console.log(tesla.getPrice(), tesla.getDescription());

// +autopilot

let tesla1 = new Tesla();
tesla1 = new Autopilot(tesla1);

console.log(tesla1.getPrice(), tesla1.getDescription());


// ---
class Audi extends Car {
    constructor() {
        super();
        this.price = 20000;
        this.model = 'Audi';
    }
}

let audi = new Audi();
audi = new Autopilot(audi);

console.log(audi.getPrice(), audi.getDescription());

File: patterns/structural_decorator/structural_decorator_webref.py
class Car:
  def __init__(self):
    self.price = 100
    self.model = 'car'

  def get_price(self):
    return self.price

  def get_description(self):
    return self.model


class Tesla(Car):
  def __init__(self):
    super()
    self.price = 500
    self.model = 'tesla'

# decorators
class Autopilot:
  def __init__(self, car):
    self.car = car

  def get_price(self):
    return self.car.get_price() + 1001

  def get_description(self):
    return (f"{self.car.get_description()} with autopilot")


if __name__ == "__main__":
  tesla = Tesla()

  tesla = Autopilot(tesla)
  print(f"{tesla.get_description()}. Price: {tesla.get_price()}")
